### DISCUSSION TITLE
API and Budibase DB Table

### DISCUSSION THREAD

**Discussion by ctroche28** [Budibase/budibase#7159] [Help]
Hi,

Is it possible to display API data into the same table as a Budibase DB table? If so, how do I go about this?

Thanks for any help or advice! I appreciate it.

**Comment by melohagan** [comment]
Hi @ctroche28 

You can use a REST query as a datasource for a table component: https://docs.budibase.com/docs/rest-queries

If you want to save the response of an API call into a Budibase DB then you could try chaining actions: https://docs.budibase.com/docs/referencing
- Execute Query
- Save Row (response from execute query)

You may need multiple save row actions if you need to save more than one row.

If the number of rows to save is unknown, then you may need to use an automation instead. It supports looping: https://docs.budibase.com/docs/automation-actions#looping

Then you could use the **External Data Connector** to make your API call, and then in your **Create Row** action you can loop over the response to save a row for each item returned by the API.

**Comment by acseven** [comment]
Hi,
I was looking for some help on storing a REST API query result on a local database. 

I've tested this somewhat successfully with your pointers on the message above, but I'm failing to understand how (if possible) can I have the data be updated whenever imported. 

I opted for an automation, and with the logic below I can get each item that results from the REST API query to be stored as a JSON field (JSON column in an internal database) This is triggered as an action with a Button.

![image](https://user-images.githubusercontent.com/1176371/192015307-e6d14660-314f-4f1f-8bbc-5e4fc7bca2c2.png)

So my question is: would it be possible to check if a given key/value (i.e. "id") is already present in the table and either update or create the row on the internal database?

Also, in order to loop through the entire database, can you suggest a way of getting the total amount of rows that an API can return? 

Thank you for any input, I appreciate it.

**Comment by acseven** [comment]
Also, I'm using a JSON field just because I couldn't make the automation process the JSON output into columns, and store those in a database with multiple columns, but that would be the best approach for my use case. Can it be done as well? 

EDIT: forget about this part, got there, probably was typing something wrong:

![image](https://user-images.githubusercontent.com/1176371/192026235-b89101f6-2b94-4b51-82df-1a65a54622a7.png)

I found one issue with this though: one of the fields should be a Boolean data type, but if I select that there's no way to populate the current item:

![image](https://user-images.githubusercontent.com/1176371/192026879-3b81534c-1acd-4b18-a9ed-1bb29a008bc6.png)

I can only import the field as text

**Reply by melohagan** [reply]
Hey @acseven Were you able to toggle the *Use values* dropdown to *Use bindings*. Did that help?

**Comment by acseven** [comment]
Thanks for the pointer; selecting "Use Bindings" does allow the field to be properly populated:

![image](https://user-images.githubusercontent.com/1176371/192241417-efa0ed5c-f329-4a94-bc5d-13278a08e8a6.png)

Nice. :-)

Could you help out understanding if these are possible and some hints on to implement them?

1. Would it be possible to check if a given key/value (i.e. "id") is already present in the table and either update or create the row on the internal database? I though of looping an Update Row action first which should work for existing records, and then running Create row for those that are not but for this I can't find a way of filtering out entries that already exist.

2. Also, in order to loop through the entire database, can you suggest a way of getting the total amount of rows that an API can return?

3. Using a button to trigger the automation, we don't get any actual feedback on the process, just that it has been initiated; in this regard I tried to find a way of adding a "working, please" indicator that stopped when the loop ended, but couldn't find none. Is there any way of doing this? Ideally, some actually progress indication ("phase 1: x out y records looped") would also be helpful, but a "working" indicator would suffice.

**Reply by melohagan** [reply]
Hey @acseven 

1. This may help: https://github.com/Budibase/budibase/discussions/6294
2. I believe the total number of rows that can be returned by an API call is 1000 on cloud. Otherwise you could check the length of the response array?
3. The automations run asynchronously on a worker thread. There's no communication between automations and the frontend so there's no way to do this. You can check the automation logs if you just need to verify if an automation was successful: https://docs.budibase.com/docs/logs
